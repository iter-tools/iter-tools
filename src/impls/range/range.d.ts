import { IterableIterator } from '../../types/iterable';

declare function range(end: number): IterableIterator<number>;
declare function range(start: number, end: number, step?: number): IterableIterator<number>;
declare function range(opts: {
  start?: number;
  end?: number;
  step?: number;
}): IterableIterator<number>;
declare function range(_?: undefined): IterableIterator<number>;

export { range };
