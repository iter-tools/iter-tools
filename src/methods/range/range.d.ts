import { GeneratorIterator } from '../../internal/iterable';

declare function range(end: number): GeneratorIterator<number>;
declare function range(start: number, end: number, step?: number): GeneratorIterator<number>;
declare function range(opts: {
  start?: number;
  end?: number;
  step?: number;
}): GeneratorIterator<number>;
declare function range(_?: undefined): GeneratorIterator<number>;

export default range;
