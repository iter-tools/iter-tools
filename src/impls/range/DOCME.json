{
  "type": "create",
  "signatures": [
    [{ "name": "start" }, { "name": "end" }, { "name": "step", "isOptional": true }],
    [{ "name": "end", "isOptional": true }],
    [{ "name": null, "properties": ["start", "end", "step"] }]
  ]
}
