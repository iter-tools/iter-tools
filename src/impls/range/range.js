import { wrapWithResultIterable } from '../../internal/iterable.js';
import { isObject, isDef } from '../../internal/shapes.js';
import { isInteger, isIntegerOrInfinite } from '../../internal/number.js';

export function* range(start = 0, end = Infinity, step = 1) {
  for (let i = start; step > 0 ? i < end : i > end; i += step) {
    yield i;
  }
}

export default wrapWithResultIterable(range, {
  validateArgs(args) {
    let [optsOrEndOrStart, end = Infinity, step] = args;
    let start;

    if (isObject(optsOrEndOrStart)) {
      ({ start, end, step } = optsOrEndOrStart);
    } else {
      if (args.length > 1) {
        start = optsOrEndOrStart;
      } else {
        end = optsOrEndOrStart;
      }
    }

    if (isDef(start) && !isInteger(start)) {
      throw new TypeError('The specified start was not an integer');
    }

    if (isDef(end) && !isIntegerOrInfinite(end)) {
      throw new TypeError('The specified end was not an integer or infinite');
    }

    if (isDef(step) && !isInteger(step, true)) {
      throw new TypeError('The specified step was not a non-zero integer');
    }

    args[0] = start;
    args[1] = end;
    args[2] = step;
  },
});
