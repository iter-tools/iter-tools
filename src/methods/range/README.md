Create an iterable returning a sequence of numbers (the sequence can be infinite).
Overloads are: `range({start, step, end})`, `range(end)`, or `range(start, step, end)`.
```js
range(); // 0, 1, 2 ... Infinity

range({end: 3}); // 0, 1, 2
range({start: 3}); // 3, 4, 5 ... Infinity
range({start: 3, end: 6}); // 3, 4, 5
range({start: 3, end: 10, step: 3}); // 3, 6, 9
range({start: 9, end: 3, step: -3}); // 9, 6

range(3); // 0, 1, 2
range(3, 6); // 3, 4, 5
range(3, 10, 3); // 3, 6, 9
```
