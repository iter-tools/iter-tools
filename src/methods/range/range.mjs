import { wrapWithMethodIterable } from '../../internal/iterable';

export function* range(start, end, step = 1) {
  for (let i = start; step > 0 ? i < end : i > end; i += step) {
    yield i;
  }
}

export default wrapWithMethodIterable(range, {
  validateArgs(args) {
    let [optsOrEndOrStart, end = Infinity, step = 1] = args;
    let start = 0;

    if (typeof optsOrEndOrStart === 'number') {
      if (args.length > 1) {
        start = optsOrEndOrStart;
      } else {
        end = optsOrEndOrStart;
      }
    } else if (optsOrEndOrStart && typeof optsOrEndOrStart === 'object') {
      ({ start = 0, end = Infinity, step = 1 } = optsOrEndOrStart);
    }

    if (typeof start !== 'number') {
      throw new TypeError('The specified start was not a number');
    }

    if (typeof end !== 'number') {
      throw new TypeError('The specified end was not a number');
    }

    if (typeof step !== 'number' || step === 0) {
      throw new TypeError('The specified step was not a number !== 0');
    }

    args[0] = start;
    args[1] = end;
    args[2] = step;
  },
});
