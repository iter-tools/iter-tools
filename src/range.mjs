export default function * range (opts) {
  let start, step, end
  opts = typeof opts === 'number' ? { end: opts, start: 0 }
    : (typeof opts === 'object' ? opts : {})
  step = typeof opts.step === 'undefined' ? 1 : opts.step
  end = typeof opts.end === 'undefined'
    ? (step > 0 ? Infinity : -Infinity) : opts.end
  start = opts.start ? opts.start : 0

  for (let i = start; step > 0 ? i < end : i > end; i += step) {
    yield i
  }
}
