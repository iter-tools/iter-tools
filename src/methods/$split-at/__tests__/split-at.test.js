/**
 * @generated-from ./$split-at.test.js
 * This file is autogenerated from a template. Please do not edit it directly.
 * To rebuild it from its template use the command
 * > npm run generate
 * More information can be found in CONTRIBUTING.md
 */

/* eslint-disable no-unused-vars,import/no-duplicates,no-constant-condition */

import { splitAt, toArray } from '../../..';
import { range } from '../../../__tests__/range';
describe('splitAt', () => {
  describe('with 0 index', () => {
    it('when all values are in second part', () => {
      const [[...first], [...second]] = splitAt(0, range(0, 6));
      expect([first, second]).toEqual([[], [0, 1, 2, 3, 4, 5]]);
    });
  });
  describe('with positive index', () => {
    it('works when the halves are consumed in order', () => {
      const [[...first], [...second]] = splitAt(3, range(0, 6));
      expect([first, second]).toEqual([[0, 1, 2], [3, 4, 5]]);
    });
    it('works when the source is exhuasted while the first half is being consumed', () => {
      const [[...first], [...second]] = splitAt(3, range(0, 2));
      expect([first, second]).toEqual([[0, 1], []]);
    });
    it('works when the source is exhuasted while the second half is being consumed', () => {
      const [[...first], [...second]] = splitAt(3, range(0, 4));
      expect([first, second]).toEqual([[0, 1, 2], [3]]);
    });
  });
  describe('with negative index', () => {
    it('works when the halves are consumed in order', () => {
      const [[...first], [...second]] = splitAt(-3, range(0, 6));
      expect([first, second]).toEqual([[0, 1, 2], [3, 4, 5]]);
    });
    it('all values are in the first part when |index| is larger than source size', () => {
      const [[...first], [...second]] = splitAt(-3, range(0, 2));
      expect([first, second]).toEqual([[0, 1], []]);
    });
  });
  describe('given strings', () => {
    describe('with 0 index', () => {
      it('when all values are in second part', () => {
        const [first, second] = splitAt(0, 'abcdef');
        expect([first, second]).toEqual(['', 'abcdef']);
      });
    });
    describe('with positive index', () => {
      it('works when the string is longer than the split index', () => {
        const [first, second] = splitAt(3, 'abcdef');
        expect([first, second]).toEqual(['abc', 'def']);
      });
      it('works when the string is shorter than the split index', () => {
        const [first, second] = splitAt(3, 'ab');
        expect([first, second]).toEqual(['ab', '']);
      });
    });
    describe('with negative index', () => {
      it('works when the string is longer than the split index', () => {
        const [first, second] = splitAt(-3, 'abcdef');
        expect([first, second]).toEqual(['abc', 'def']);
      });
      it('works when the string is shorter than the split index', () => {
        const [first, second] = splitAt(-3, 'ab');
        expect([first, second]).toEqual(['', 'ab']);
      });
    });
  });
  describe('with positive index', () => {
    it('works when the halves are consumed in order', () => {
      const [first, second] = splitAt(3, range(0, 6));
      expect([toArray(first), toArray(second)]).toEqual([[0, 1, 2], [3, 4, 5]]);
    });
    it('works when the source is exhuasted while the first half is being consumed', () => {
      const [first, second] = splitAt(3, range(0, 2));
      expect([toArray(first), toArray(second)]).toEqual([[0, 1], []]);
    });
    it('works when the source is exhuasted while the second half is being consumed', () => {
      const [first, second] = splitAt(3, range(0, 4));
      expect([toArray(first), toArray(second)]).toEqual([[0, 1, 2], [3]]);
    });
  });
  describe('with negative index', () => {
    it('works when the halves are consumed in order', () => {
      const [first, second] = splitAt(-3, range(0, 6));
      expect([toArray(first), toArray(second)]).toEqual([[0, 1, 2], [3, 4, 5]]);
    });
    it('all values are in the first part when |index| is larger than source size', () => {
      const [first, second] = splitAt(-3, range(0, 2));
      expect([toArray(first), toArray(second)]).toEqual([[0, 1], []]);
    });
  });
  it('allows only the second half to being consumed', () => {
    const [, second] = splitAt(3, range(0, 6));
    expect(toArray(second)).toEqual([3, 4, 5]);
  });
  it('throws if only the first half is taken', () => {
    let error;

    try {
      const [first] = splitAt(3, range(0, 6));
      toArray(first);
    } catch (e) {
      error = e;
    }

    expect(error).toMatchSnapshot();
  });
  it('throws when the second half is consumed before the first', () => {
    const [first, second] = splitAt(3, range(0, 6));
    expect(toArray(second)).toEqual([3, 4, 5]);
    let error;

    try {
      toArray(first);
    } catch (e) {
      error = e;
    }

    expect(error).toMatchSnapshot();
  });
});
