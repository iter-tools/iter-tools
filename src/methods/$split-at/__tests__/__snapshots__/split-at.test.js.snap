// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitAt throws if only the first half is taken 1`] = `[Error: You must take both parts of a bisector or neither.]`;

exports[`splitAt throws when the second half is consumed before the first 1`] = `[Error: Cannot take from this split part. It is no longer the active part.]`;
